services/docker/docker-version.service.js

```diff
@@ -0,0 +1,102 @@
+'use strict'
+
+const Joi = require('@hapi/joi')
+const { anyInteger } = require('../validators')
+const {
+  buildDockerUrl,
+  getDockerHubUser,
+  getMultiPageData,
+} = require('./docker-helpers')
+const { BaseJsonService } = require('..')
+const { NotFound } = require('..')
+
+const buildSchema = Joi.object({
+  count: anyInteger,
+  results: Joi.array()
+    .items(
+      Joi.object({
+        name: Joi.string(),
+        images: Joi.array().items(
+          Joi.object({
+            digest: Joi.string(),
+            architecture: Joi.string(),
+          })
+        ),
+      }).required()
+    )
+    .required(),
+}).required()
+
+module.exports = class DockerVersion extends BaseJsonService {
+  static get category() {
+    return 'version'
+  }
+
+  static get route() {
+    return buildDockerUrl('version', true)
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Docker Image Version',
+        pattern: ':user/:repo',
+        namedParams: { user: '_', repo: 'alpine' },
+        staticPreview: this.render({ version: '3.11.2' }),
+      },
+      {
+        title: 'Docker Image Version (tag)',
+        pattern: ':user/:repo/:tag',
+        namedParams: { user: '_', repo: 'alpine', tag: '3.10' },
+        staticPreview: this.render({ version: '3.10.3' }),
+      },
+    ]
+  }
+
+  static get defaultBadgeData() {
+    return { label: 'version' }
+  }
+
+  static render({ version }) {
+    return {
+      message: version,
+      color: 'blue',
+    }
+  }
+
+  async fetch({ user, repo, page }) {
+    page = (page && `&page=${page}`) || ''
+    return this._requestJson({
+      schema: buildSchema,
+      url: `https://registry.hub.docker.com/v2/repositories/${getDockerHubUser(
+        user
+      )}/${repo}/tags?page_size=100&ordering=last_updated${page}`,
+    })
+  }
+
+  async handle({ user, repo, tag = 'latest' }) {
```

Comment:
Should we be explicitly setting a tag value, given that we're using the `ordering=last_updated` param and letting the registry tell us what the most recent tag is? Are we sure the `latest` tag will always be identical the most recently updated tag? 

Haven't actually played with the tags in the response, but if I had published `v1.0.0` and `latest` tags yesterday, and then published `v2.0.0` and a `v3.0.0` tags I'd expect the badge to show `v3.0.0`

Also since the registry is sorting the tags for us, do we need to do any of the below searching/filtering in instances where the user hasn't specified an explicit tag? Any reason we couldn't just use `data[0].name`?

Reply:
It's Docker convention for the latest stable tag to match `latest`. If you try to pull a container without any tags at all, it defaults to latest.
This explains why we are doing the look up on it's `digest` to determine the latest SemVer that we use to display in the badge.

An example is `_/alpine` if you do not provide a tag we will lookup `latest`, this will point you to `3.11.3` which is the latest stable.
If you look up `3.11` the result also returns `3.11.3`, if you however look up `3.10`, it returns `3.10.4`.

We are utilising the the `ordering` method purely just to order the results, we could remove it, but it makes more sense for the latest tags to be at the top of the result set because it's less likely someone will look up older tags, however still possible.

Reply:
I know that the docker client and friends will default to the `latest` tag in the absence of one being specified, but that doesn't mean the owner of every single image repo is creating and publishing a latest tag every time.

We could easily support both patterns and save some overhead by still supporting the latest tag mapping when Shields users specify the latest tag (`/<<owner>>/<<repo>>/latest`)
like
https://shields-staging-pr-4562.herokuapp.com/docker/version/_/alpine/latest
![](https://shields-staging-pr-4562.herokuapp.com/docker/version/_/alpine/latest)

but on (`/<<owner>>/<<repo>>`) we can just use the tag the registry says was the most recently updated, avoid making the paginated API calls, and then avoid doing all the processing work you have below

https://shields-staging-pr-4562.herokuapp.com/docker/version/_/alpine
![](https://shields-staging-pr-4562.herokuapp.com/docker/version/_/alpine)

If we default this to `latest` always, then you prevent image authors that have an old `latest` tag from being able to display their most recent version

Reply:
I would argue that if someone is forgetting to tag their latest stable with `latest` then they're doing something wrong.

If we go down the route that you have suggested this means that `development` and potentially unsafe images could be the default badge that's shown when no tag is provided, as opposed to the safest image (read: stable) which I think is what we should be associating with `latest`.

For example many repos have `master`, `development` or `nightly` tags which are pushed to far more often than their latest, I personally would not intentionally want this to be shown by default because of the ramifications it may have. Also for most people that have been utilising this feature with the MicroBadger badge, this is how it works today and I think it's quite a sane default.

Regarding your edit, for an image like `node` if we do not parse additional pages we cannot get the latest SemVer value. When I was looking at the dataset for said image even with the sorting, the latest SemVer was on either the third or second page.

I can change it if you like but I really don't think it's the right approach because of the aforementioned reasons. Please let me know.

Reply:
> Regarding your edit, for an image like node if we do not parse additional pages we cannot get the latest SemVer value. When I was looking at the dataset for said image even with the sorting, the latest SemVer was on either the third or second page.

SemVer isn't needed when a user includes the `?sort=date` query param on their badge. Sorry that wasn't clear in my original post, getting late my time 😄 
> I would argue that if someone is forgetting to tag their latest stable with latest then they're doing something wrong.
> If we go down the route that you have suggested this means that development and potentially unsafe images could be the default badge that's shown when no tag is provided, as opposed to the safest image (read: stable) which I think is what we should be associating with latest

The point is that it's up to users, and there's no need for Shields to remove that selection from users, even if that's how you prefer to manage your images/is common in your experience. 

Users that want `latest` can get it, users that want the version badge based on date or semver (regardless of if/how they choose to use their latest tag) can get it too.

Reply:
Case in point:

Reply:
Here's the behavior I think we should be targetting (assumes the addition of a `sort` query param see [GH Tag Service for reference](https://github.com/badges/shields/blob/master/services/github/github-tag.service.js#L12-L17))

* `/v/user/repo?sort=date` - use the version of the first tag result provided by the registry in the response, and skip the paginated calls (unless the tag is `latest`, in which case make the additional calls and then map to the prettier version text)
* `/v/user/repo?sort=semver` - make the paginated calls, and use the highest semver version found
    * if there's no semver compliant values, but there is a `latest` tag, then use whatever tag latest is mapped to
    * if there's no semver compliant values, and no `latest` tag, then return the most recent value as reported in the registry response
* `/v/user/repo/tag?sort=date` - same as the tagless sort above, just with the filtering mechanism on the tags (and if there's no matching tags found then throw a descriptive error like `NotFound({ prettyMessage: 'no tags found' })`)
* `/v/user/repo/tag?sort=semver` - same as the tagless sort above, just with the filtering mechanism on the tags (and if there's no matching tags found then throw a descriptive error like `NotFound({ prettyMessage: 'no tags found' })`)

Reply:
`/v/user/repo/tag?sort=date` can't be the same as `/v/user/repo?sort=date` because the filtering mechanism on the tag requires the larger data set to perform the matching.

I'll give this a try and see what you think.

Reply:
After thinking about this even more, I think only allowing the sort params when no tag is provided probably makes the most sense.

Let me know if you want me to adjust this.